#!/bin/bash
#SBATCH --time=168:00:00
#SBATCH -p hbfraser
#SBATCH --mem=64GB
#SBATCH -c 1

cat *fasta.out > PrimatesAnc007_Merged_rmsk.fasta.out

#Need to remove repeat sequences annotated as structural RNA copies, satellites, tandem repeats, and low complexity annotations
python ../filter_repeats.py PrimatesAnc007_Merged_rmsk.fasta.out
awk -v OFS='\t' '{print $5,$6,$7}' PrimatesAnc007_Merged_rmsk.fasta.filt.out | sort -k1,1 -k2,2n | bedtools merge -i - > PrimatesAnc007_Merged_rmsk.fasta.filt.bed

../cactus-bin-v2.6.13/bin/halLiftover ../hg38.447way.hal PrimatesAnc007 PrimatesAnc007_Merged_rmsk.fasta.filt.bed PrimatesAnc015 PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc015.bed --noDupes --outPSL
../cactus-bin-v2.6.13/bin/halLiftover ../hg38.447way.hal PrimatesAnc007 PrimatesAnc007_Merged_rmsk.fasta.filt.bed PrimatesAnc024 PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc024.bed --noDupes --outPSL
../cactus-bin-v2.6.13/bin/halLiftover ../hg38.447way.hal PrimatesAnc007 PrimatesAnc007_Merged_rmsk.fasta.filt.bed PrimatesAnc040 PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc040.bed --noDupes --outPSL
../cactus-bin-v2.6.13/bin/halLiftover ../hg38.447way.hal PrimatesAnc007 PrimatesAnc007_Merged_rmsk.fasta.filt.bed PrimatesAnc041 PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc041.bed --noDupes --outPSL

awk -v OFS='\t' '{print $10,$12,$13,$14,$16,$17}' PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc015.bed > PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc015.bed"_temp" && mv PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc015.bed"_temp" PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc015.bed
bedtools intersect -a PrimatesAnc007_Merged_rmsk.fasta.filt.bed -b PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc015.bed > PrimatesAnc007_Merged_rmsk.fasta.filt.bed"_temp" && mv PrimatesAnc007_Merged_rmsk.fasta.filt.bed"_temp" PrimatesAnc007_Merged_rmsk.fasta.filt.bed

awk -v OFS='\t' '{print $10,$12,$13,$14,$16,$17}' PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc024.bed > PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc024.bed"_temp" && mv PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc024.bed"_temp" PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc024.bed
bedtools intersect -a PrimatesAnc007_Merged_rmsk.fasta.filt.bed -b PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc024.bed > PrimatesAnc007_Merged_rmsk.fasta.filt.bed"_temp" && mv PrimatesAnc007_Merged_rmsk.fasta.filt.bed"_temp" PrimatesAnc007_Merged_rmsk.fasta.filt.bed

awk -v OFS='\t' '{print $10,$12,$13,$14,$16,$17}' PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc040.bed > PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc040.bed"_temp" && mv PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc040.bed"_temp" PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc040.bed
bedtools intersect -a PrimatesAnc007_Merged_rmsk.fasta.filt.bed -b PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc040.bed > PrimatesAnc007_Merged_rmsk.fasta.filt.bed"_temp" && mv PrimatesAnc007_Merged_rmsk.fasta.filt.bed"_temp" PrimatesAnc007_Merged_rmsk.fasta.filt.bed

awk -v OFS='\t' '{print $10,$12,$13,$14,$16,$17}' PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc041.bed > PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc041.bed"_temp" && mv PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc041.bed"_temp" PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc041.bed
bedtools intersect -a PrimatesAnc007_Merged_rmsk.fasta.filt.bed -b PrimatesAnc007_Merged_rmsk.fasta.filt.PrimatesAnc041.bed > PrimatesAnc007_Merged_rmsk.fasta.filt.bed"_temp" && mv PrimatesAnc007_Merged_rmsk.fasta.filt.bed"_temp" PrimatesAnc007_Merged_rmsk.fasta.filt.bed

bedops --chop 1 PrimatesAnc007_Merged_rmsk.fasta.filt.bed | shuf -n 100000 | sort -k1,1 -k2,2n | bedtools merge -i - > PrimatesAnc007_Merged_rmsk.fasta.filt.100kb.bed
#For the next step, we print out which species we need using a short python script
#python ../get_subtree_species.py ../hg38.447way.tree.nh PrimatesAnc007

../cactus-bin-v2.6.13/bin/hal2maf --refTargets PrimatesAnc007_Merged_rmsk.fasta.filt.100kb.bed --refGenome PrimatesAnc007 --onlyOrthologs --targetGenomes "Papio_anubis,Papio_papio,Papio_hamadryas,Papio_ursinus,Papio_kindae,Papio_cynocephalus,Lophocebus_aterrimus,Theropithecus_gelada,Cercocebus_atys,Cercocebus_lunulatus,Cercocebus_torquatus,Cercocebus_chrysogaster,Mandrillus_leucophaeus,Mandrillus_sphinx,Macaca_leonina,Macaca_silenus,Macaca_siberu,Macaca_nemestrina,Macaca_nigra,Macaca_maura,Macaca_tonkeana,Macaca_fuscata,Macaca_cyclopis,Macaca_mulatta,Macaca_fascicularis,Macaca_assamensis,Macaca_thibetana,Macaca_radiata,Macaca_arctoides,Cercopithecus_albogularis,Cercopithecus_nictitans,Cercopithecus_ascanius,Cercopithecus_cephus,Cercopithecus_petaurista,Cercopithecus_lowei,Cercopithecus_mona,Cercopithecus_pogonias,Cercopithecus_neglectus,Cercopithecus_diana,Cercopithecus_roloway,Cercopithecus_hamlyni,Miopithecus_ogouensis,Allochrocebus_preussi,Allochrocebus_lhoesti,Allochrocebus_solatus,Chlorocebus_aethiops,Chlorocebus_pygerythrus,Chlorocebus_sabaeus,Erythrocebus_patas,Allenopithecus_nigroviridis,Trachypithecus_phayrei,Trachypithecus_crepusculus,Trachypithecus_obscurus,Trachypithecus_melamera,Trachypithecus_auratus,Trachypithecus_cristatus,Trachypithecus_germaini,Trachypithecus_francoisi,Trachypithecus_leucocephalus,Trachypithecus_laotum,Trachypithecus_hatinhensis,Trachypithecus_geei,Trachypithecus_pileatus,Semnopithecus_entellus,Semnopithecus_hypoleucos,Semnopithecus_priam,Semnopithecus_schistaceus,Semnopithecus_johnii,Semnopithecus_vetulus,Pygathrix_cinerea,Pygathrix_nigripes_b,Pygathrix_nigripes_a,Rhinopithecus_strykeri,Rhinopithecus_bieti,Rhinopithecus_roxellana,Nasalis_larvatus,Presbytis_mitrata,Presbytis_comata,Colobus_polykomos,Colobus_guereza,Colobus_angolensis,Piliocolobus_tephrosceles,Piliocolobus_gordonorum,Piliocolobus_kirkii,Piliocolobus_badius,Pan_paniscus,Pan_troglodytes,Homo_sapiens,Gorilla_gorilla,Gorilla_beringei,Pongo_abelii,Pongo_pygmaeus,Hylobates_pileatus_b,Hylobates_pileatus_a,Hylobates_abbotti,Hylobates_muelleri,Hylobates_agilis,Hylobates_klossii,Hoolock_leuconedys,Nomascus_siki_b,Nomascus_siki_a,Nomascus_gabriellae,Nomascus_annamensis,Nomascus_concolor,Symphalangus_syndactylus" ../hg38.447way.hal PrimatesAnc007.100kb.maf

#The values produced from this command line up nicely with existing values from the all mammals alignment!
../phast/bin/phyloFit  PrimatesAnc007.100kb.maf -i MAF --subst-mod REV --EM --tree ../hg38.447way.scientificNames.nh.txt --out-root PrimatesAnc007.100kb

../cactus-bin-v2.6.13/bin/halLiftover ../hg38.447way.hal PrimatesAnc007 PrimatesAnc007_Merged_rmsk.fasta.filt.bed Homo_sapiens PrimatesAnc007_Merged_rmsk.fasta.filt.Homo_sapiens.bed --noDupes --outPSL

#Now we will do it for chrX and chrY
awk -v OFS='\t' '{if ($14=="chrX") print $10,$12,$13,$14,$16,$17}' PrimatesAnc007_Merged_rmsk.fasta.filt.Homo_sapiens.bed | bedops --chop 1 - | shuf -n 100000 | sort -k1,1 -k2,2n | bedtools merge -i - > PrimatesAnc007_Merged_rmsk.fasta.filt.chrX.100kb.bed
awk -v OFS='\t' '{if ($14=="chrY") print $10,$12,$13,$14,$16,$17}' PrimatesAnc007_Merged_rmsk.fasta.filt.Homo_sapiens.bed | bedops --chop 1 - | shuf -n 100000 | sort -k1,1 -k2,2n | bedtools merge -i - > PrimatesAnc007_Merged_rmsk.fasta.filt.chrY.100kb.bed

../cactus-bin-v2.6.13/bin/hal2maf --refTargets PrimatesAnc007_Merged_rmsk.fasta.filt.chrX.100kb.bed --refGenome PrimatesAnc007 --onlyOrthologs --targetGenomes "Papio_anubis,Papio_papio,Papio_hamadryas,Papio_ursinus,Papio_kindae,Papio_cynocephalus,Lophocebus_aterrimus,Theropithecus_gelada,Cercocebus_atys,Cercocebus_lunulatus,Cercocebus_torquatus,Cercocebus_chrysogaster,Mandrillus_leucophaeus,Mandrillus_sphinx,Macaca_leonina,Macaca_silenus,Macaca_siberu,Macaca_nemestrina,Macaca_nigra,Macaca_maura,Macaca_tonkeana,Macaca_fuscata,Macaca_cyclopis,Macaca_mulatta,Macaca_fascicularis,Macaca_assamensis,Macaca_thibetana,Macaca_radiata,Macaca_arctoides,Cercopithecus_albogularis,Cercopithecus_nictitans,Cercopithecus_ascanius,Cercopithecus_cephus,Cercopithecus_petaurista,Cercopithecus_lowei,Cercopithecus_mona,Cercopithecus_pogonias,Cercopithecus_neglectus,Cercopithecus_diana,Cercopithecus_roloway,Cercopithecus_hamlyni,Miopithecus_ogouensis,Allochrocebus_preussi,Allochrocebus_lhoesti,Allochrocebus_solatus,Chlorocebus_aethiops,Chlorocebus_pygerythrus,Chlorocebus_sabaeus,Erythrocebus_patas,Allenopithecus_nigroviridis,Trachypithecus_phayrei,Trachypithecus_crepusculus,Trachypithecus_obscurus,Trachypithecus_melamera,Trachypithecus_auratus,Trachypithecus_cristatus,Trachypithecus_germaini,Trachypithecus_francoisi,Trachypithecus_leucocephalus,Trachypithecus_laotum,Trachypithecus_hatinhensis,Trachypithecus_geei,Trachypithecus_pileatus,Semnopithecus_entellus,Semnopithecus_hypoleucos,Semnopithecus_priam,Semnopithecus_schistaceus,Semnopithecus_johnii,Semnopithecus_vetulus,Pygathrix_cinerea,Pygathrix_nigripes_b,Pygathrix_nigripes_a,Rhinopithecus_strykeri,Rhinopithecus_bieti,Rhinopithecus_roxellana,Nasalis_larvatus,Presbytis_mitrata,Presbytis_comata,Colobus_polykomos,Colobus_guereza,Colobus_angolensis,Piliocolobus_tephrosceles,Piliocolobus_gordonorum,Piliocolobus_kirkii,Piliocolobus_badius,Pan_paniscus,Pan_troglodytes,Homo_sapiens,Gorilla_gorilla,Gorilla_beringei,Pongo_abelii,Pongo_pygmaeus,Hylobates_pileatus_b,Hylobates_pileatus_a,Hylobates_abbotti,Hylobates_muelleri,Hylobates_agilis,Hylobates_klossii,Hoolock_leuconedys,Nomascus_siki_b,Nomascus_siki_a,Nomascus_gabriellae,Nomascus_annamensis,Nomascus_concolor,Symphalangus_syndactylus" ../hg38.447way.hal PrimatesAnc007.100kb.chrX.maf
../cactus-bin-v2.6.13/bin/hal2maf --refTargets PrimatesAnc007_Merged_rmsk.fasta.filt.chrY.100kb.bed --refGenome PrimatesAnc007 --onlyOrthologs --targetGenomes "Papio_anubis,Papio_papio,Papio_hamadryas,Papio_ursinus,Papio_kindae,Papio_cynocephalus,Lophocebus_aterrimus,Theropithecus_gelada,Cercocebus_atys,Cercocebus_lunulatus,Cercocebus_torquatus,Cercocebus_chrysogaster,Mandrillus_leucophaeus,Mandrillus_sphinx,Macaca_leonina,Macaca_silenus,Macaca_siberu,Macaca_nemestrina,Macaca_nigra,Macaca_maura,Macaca_tonkeana,Macaca_fuscata,Macaca_cyclopis,Macaca_mulatta,Macaca_fascicularis,Macaca_assamensis,Macaca_thibetana,Macaca_radiata,Macaca_arctoides,Cercopithecus_albogularis,Cercopithecus_nictitans,Cercopithecus_ascanius,Cercopithecus_cephus,Cercopithecus_petaurista,Cercopithecus_lowei,Cercopithecus_mona,Cercopithecus_pogonias,Cercopithecus_neglectus,Cercopithecus_diana,Cercopithecus_roloway,Cercopithecus_hamlyni,Miopithecus_ogouensis,Allochrocebus_preussi,Allochrocebus_lhoesti,Allochrocebus_solatus,Chlorocebus_aethiops,Chlorocebus_pygerythrus,Chlorocebus_sabaeus,Erythrocebus_patas,Allenopithecus_nigroviridis,Trachypithecus_phayrei,Trachypithecus_crepusculus,Trachypithecus_obscurus,Trachypithecus_melamera,Trachypithecus_auratus,Trachypithecus_cristatus,Trachypithecus_germaini,Trachypithecus_francoisi,Trachypithecus_leucocephalus,Trachypithecus_laotum,Trachypithecus_hatinhensis,Trachypithecus_geei,Trachypithecus_pileatus,Semnopithecus_entellus,Semnopithecus_hypoleucos,Semnopithecus_priam,Semnopithecus_schistaceus,Semnopithecus_johnii,Semnopithecus_vetulus,Pygathrix_cinerea,Pygathrix_nigripes_b,Pygathrix_nigripes_a,Rhinopithecus_strykeri,Rhinopithecus_bieti,Rhinopithecus_roxellana,Nasalis_larvatus,Presbytis_mitrata,Presbytis_comata,Colobus_polykomos,Colobus_guereza,Colobus_angolensis,Piliocolobus_tephrosceles,Piliocolobus_gordonorum,Piliocolobus_kirkii,Piliocolobus_badius,Pan_paniscus,Pan_troglodytes,Homo_sapiens,Gorilla_gorilla,Gorilla_beringei,Pongo_abelii,Pongo_pygmaeus,Hylobates_pileatus_b,Hylobates_pileatus_a,Hylobates_abbotti,Hylobates_muelleri,Hylobates_agilis,Hylobates_klossii,Hoolock_leuconedys,Nomascus_siki_b,Nomascus_siki_a,Nomascus_gabriellae,Nomascus_annamensis,Nomascus_concolor,Symphalangus_syndactylus" ../hg38.447way.hal PrimatesAnc007.100kb.chrY.maf

../phast/bin/phyloFit PrimatesAnc007.100kb.chrX.maf -i MAF --subst-mod REV --EM --tree ../hg38.447way.scientificNames.nh.txt --out-root PrimatesAnc007.100kb.chrX
../phast/bin/phyloFit PrimatesAnc007.100kb.chrY.maf -i MAF --subst-mod REV --EM --tree ../hg38.447way.scientificNames.nh.txt --out-root PrimatesAnc007.100kb.chrY
