{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7283576",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PosSelect_Functions_Old import *\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import gaussian_kde\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy\n",
    "import seaborn as sns\n",
    "from scipy.stats import mannwhitneyu as mwu\n",
    "from scipy.stats import ttest_ind\n",
    "from scipy.stats import ttest_rel\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from scipy.stats import wilcoxon\n",
    "from scipy.optimize import curve_fit\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats import norm\n",
    "from collections import Counter\n",
    "\n",
    "hfont = {'fontname':'Arial'}\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "\n",
    "#Code borrowed heavily from here: https://stackoverflow.com/questions/62375034/find-non-overlapping-area-between-two-kde-plots\n",
    "plt.rcParams.update(\n",
    "    {\"text.usetex\": False}\n",
    ")\n",
    "\n",
    "def parse_table(i):\n",
    "    return [int(x) for x in i.replace(\"[\", \"\").replace(\"]\", \"\").split(\",\")]\n",
    "\n",
    "d_abrev = {\"LiangSteinNeuron\":\"FC exc. neur.\", \"FetalChondrocytes\":\"F chond.\", \"SertoliMale\":\"FG sertoli\", \"preGC_IIaFemale\":\"FG preGC IIa\",\\\n",
    "          \"NeuralFemale\":\"FG neur.\", \"FetalGonadImmuneFemale\":\"FG immune\", \"VIP\":\"AC VIP inh. neur.\", \"LiangSteinProgenitor\":\"FC prog.\",\\\n",
    "          \"AdultHeartVentricularCardiomyocyte\":\"AH cardiomyo.\", \"AdultLoopOfHenle\":\"AK loop of henle\", \"FetalBrainNeurGlioblast_CB_VZ\":\"FCB glioblast\",\\\n",
    "         \"AdultProximalTubule\":\"AK prox. tub.\", \"FetalLeydigMale\":\"FG leydig\", \"SST\":\"AC SST inh neur.\", \"KosoyRoussosControlMicroglia\":\"AC microglia\",\\\n",
    "         \"FetalBrainFloorPlate\":\"FB fl. plate\", \"FetalArterialECs\":\"FH endoth.\", \"ASCT\":\"AC astro.\", \"FetalBrainCOP\":\"FB COP\",\\\n",
    "         \"AMY\":\"AA neur.\", \"PVALB\":\"AC PVALB inh neur.\", \"ITL23\":\"AC L2-3 IT neur.\", \"FetalBrainNeurCB_GNP_IPC_1\":\"FB inter. prog.\", \"FetalBrainNeurDAergic\":\"FB DA neur.\",\\\n",
    "          \"OGC\":\"AC Oligo.\", \"D1Pu\":\"AP D1 inh neur.\", \"FetalBrainNeurSerotonergic\":\"FB 5-HT neur.\", \"FetalBrainNeurDRG_2\":\"FS DRG neur.\",\\\n",
    "          \"FetalHeartPericytes\":\"FH peri.\", \"FetalHeartEndocardium\":\"FH endocard.\", \"FetalHeartCardiacFibroblasts\":\"FH fibro.\", \"FetalBrainNeurPurkinje_6\":\"FCB Purk. inh neur.\",\\\n",
    "          \"AdultHeartSmoothMuscle\":\"AH smooth musc.\", \"FetalBrainRoofPlate\":\"FB ro. plate\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cdaf544",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plotting across chromosomes using PhyloP\n",
    "df = pd.read_csv(\"PerChrom_Results/PerChrom_PhyloP447_SpecSup250_MAFCut0.25_AllChroms_Bootstrap.txt.gz\", sep = \"\\t\")\n",
    "df = df[df[\"Proportion\"] == 0.6]\n",
    "\n",
    "medians = []\n",
    "ci = []\n",
    "palette = {}\n",
    "for ct in np.unique(df[\"Chrom\"]):\n",
    "    dff = df[(df[\"Chrom\"].isin([ct]))]\n",
    "    dff = dff.sort_values(\"alpha\")\n",
    "    ci.append([ct, list(dff[\"alpha\"])[25], list(dff[\"alpha\"])[975]])\n",
    "    medians.append([ct, np.median(dff[\"alpha\"])])\n",
    "    \n",
    "df_med = pd.DataFrame(medians).sort_values(1, ascending = False)\n",
    "order = list(df_med[0])\n",
    "\n",
    "df_ci = pd.DataFrame(ci)\n",
    "df_ci.columns = [\"Chromosome\", \"Lower95\", \"Upper95\"]\n",
    "df_ci = df_ci.sort_values(\"Lower95\", ascending = False)\n",
    "\n",
    "df.columns = [\"Chromosome\"] + list(df.columns)[1:]\n",
    "\n",
    "for index, row in df_ci.iterrows():\n",
    "    if row[\"Lower95\"] > 0:\n",
    "        palette[row[\"Chromosome\"]] = \"#F42FF5\"\n",
    "    else:\n",
    "        palette[row[\"Chromosome\"]] = \"#40A94D\"\n",
    "\n",
    "fig, ax = plt.subplots(figsize = (10, 4))\n",
    "sns.boxplot(data = df, y = \"alpha\", x = \"Chromosome\", order = order, palette = palette)\n",
    "plt.ylim(-0.15, 0.075)\n",
    "plt.ylabel(\"$\\\\alpha_{Cons}$\", size = 16)\n",
    "plt.xlabel(\"Chromosome\", size = 16)\n",
    "plt.title(\"Positive selection on non-coding sites per chromosome\", size = 20)\n",
    "plt.xticks(rotation = 90, size = 18)\n",
    "#plt.title(\"Positive selection in HARs across cell types\", size = 25)\n",
    "\n",
    "df_ci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "931d2101",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Exploring the chromosome X stuff\n",
    "v, yvalls = read_noncoding_data_fast(path = \"./\", maf_cut = 0.25, spec_sup = 250)\n",
    "\n",
    "yvalls.index = yvalls[\"Position\"]\n",
    "yvalls = add_unfold(yvalls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c121ff31",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Because vv is faster to type\n",
    "vv = yvalls.copy()\n",
    "yvalls = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "840c9ae1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Restrict to chrX\n",
    "v[\"Chrom\"] = [x.split(\":\")[0] for x in v[\"Position\"]]\n",
    "vv[\"Chrom\"] = [x.split(\":\")[0] for x in vv[\"Position\"]]\n",
    "\n",
    "vx = v[v[\"Chrom\"].isin([\"chrX\"])]\n",
    "vvx = vv[vv[\"Chrom\"].isin([\"chrX\"])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07fc9299",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot chrX\n",
    "vvv = prepare_alpha(vx, vvx)\n",
    "z = [np.float64(j) for j in list(vvx[\"PhyloP447\"])]\n",
    "z.sort()\n",
    "cuttt = 0.95\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "\n",
    "alpha = compute_alpha_cutoff(vvv, dn_cut = 0.01, plot = True, cutoff = cutoff, title = \"Excess of fixed substitutions in conserved sites on chrX\")\n",
    "table2 = alpha[3]\n",
    "table1 = alpha[2]\n",
    "\n",
    "p = (fisher_exact(table1, alternative = \"greater\")[1] + fisher_exact(table2, alternative = \"greater\")[1])/2\n",
    "print(alpha, p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2290217f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "vvv = prepare_alpha(vx, vvx)\n",
    "z = [np.float64(j) for j in list(vvx[\"PhyloP447\"])]\n",
    "z.sort()\n",
    "cuttt = 0.6\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "\n",
    "alpha = compute_alpha_cutoff(vvv, dn_cut = 0.01, plot = True, cutoff = cutoff, title = \"Excess of fixed substitutions in conserved sites on chrX\", window = [3, 12])\n",
    "table2 = alpha[3]\n",
    "table1 = alpha[2]\n",
    "\n",
    "p = (fisher_exact(table1, alternative = \"greater\")[1] + fisher_exact(table2, alternative = \"greater\")[1])/2\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b459ef64",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read in the chrX selective sweep regions\n",
    "kl = pd.read_csv(\"KasperLange_CellReports2023_chrX.csv\", header = None)\n",
    "vx_kl = pd.DataFrame()\n",
    "vvx_kl = pd.DataFrame()\n",
    "\n",
    "for index, row in kl.iterrows():\n",
    "    vx_kl = pd.concat([vx_kl, vx[vx[\"Position\"].isin([\"chrX:\" + str(i) for i in range(row[1], row[2])])]])\n",
    "    vvx_kl = pd.concat([vvx_kl, vvx[vvx[\"Position\"].isin([\"chrX:\" + str(i) for i in range(row[1], row[2])])]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8ee84de",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test on those regions\n",
    "vvv = prepare_alpha(vx_kl, vvx_kl, stat = \"PhyloP447\")\n",
    "cuttt = 0.95\n",
    "z = list(vvx_kl[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Known selective sweeps\")\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d4d18e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test on those regions\n",
    "vvv = prepare_alpha(vx_kl, vvx_kl, stat = \"PhyloP447\")\n",
    "cuttt = 0.6\n",
    "z = list(vvx_kl[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Known selective sweeps\", window = [3, 12])\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)\n",
    "\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5104d55",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test on everything without those regions\n",
    "vx_nkl = vx[~vx[\"Position\"].isin(vx_kl[\"Position\"])].copy()\n",
    "vvx_nkl = vvx[~vvx[\"Position\"].isin(vvx_kl[\"Position\"])].copy()\n",
    "vvv = prepare_alpha(vx_nkl, vvx_nkl, stat = \"PhyloP447\")\n",
    "cuttt = 0.95\n",
    "z = list(vvx_nkl[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Not in known selective sweeps\")\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a543e79a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test on everything without those regions\n",
    "vx_nkl = vx[~vx[\"Position\"].isin(vx_kl[\"Position\"])].copy()\n",
    "vvx_nkl = vvx[~vvx[\"Position\"].isin(vvx_kl[\"Position\"])].copy()\n",
    "vvv = prepare_alpha(vx_nkl, vvx_nkl, stat = \"PhyloP447\")\n",
    "cuttt = 0.9\n",
    "z = list(vvx_nkl[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Not in known selective sweeps\", window = [3, 12])\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)\n",
    "\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aecc3c9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test only on those regions near our multi-copy and ampliconic expansions\n",
    "amp = pd.read_csv(\"chrX_Ampliconic.csv\")\n",
    "amp = amp[amp[\"Gene class\"].isin([\"Multicopy\", \"Ampliconic\"])]\n",
    "\n",
    "vx_amp = vx[vx[\"NearestGene\"].isin(amp[\"Gene name\"])]\n",
    "vvx_amp = vvx[vvx[\"NearestGene\"].isin(amp[\"Gene name\"])]\n",
    "\n",
    "#Test on those regions\n",
    "vvv = prepare_alpha(vx_amp, vvx_amp, stat = \"PhyloP447\")\n",
    "cuttt = 0.95\n",
    "z = list(vvx_amp[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Near known multicopy testis-expressed genes\")\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7299e508",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "188ac5f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test on those regions\n",
    "vvv = prepare_alpha(vx_amp, vvx_amp, stat = \"PhyloP447\")\n",
    "cuttt = 0.9\n",
    "z = list(vvx_amp[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Near known multicopy testis-expressed genes\", window = [3, 12])\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)\n",
    "\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8175acde",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test only on anything not in those regions near our multi-copy and ampliconic expansions\n",
    "vx_namp = vx[~vx[\"Position\"].isin(vx_amp[\"Position\"])].copy()\n",
    "vvx_namp = vvx[~vvx[\"Position\"].isin(vvx_amp[\"Position\"])].copy()\n",
    "vvv = prepare_alpha(vx_nkl, vvx_nkl, stat = \"PhyloP447\")\n",
    "cuttt = 0.9\n",
    "z = list(vvx_amp[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Not near known multicopy testis-expressed genes\")\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d900cd7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test only on anything not in those regions near our multi-copy and ampliconic expansions\n",
    "vx_namp = vx[~vx[\"Position\"].isin(vx_amp[\"Position\"])].copy()\n",
    "vvx_namp = vvx[~vvx[\"Position\"].isin(vvx_amp[\"Position\"])].copy()\n",
    "vvv = prepare_alpha(vx_nkl, vvx_nkl, stat = \"PhyloP447\")\n",
    "cuttt = 0.9\n",
    "z = list(vvx_amp[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Not near known multicopy testis-expressed genes\", window = [3, 12])\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)\n",
    "\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87ce5e06",
   "metadata": {},
   "outputs": [],
   "source": [
    "vx_use = vx[(vx[\"NearestGene\"].isin(amp[\"Gene name\"])) | (vx[\"Position\"].isin(vx_kl[\"Position\"]))]\n",
    "vvx_use = vvx[(vvx[\"NearestGene\"].isin(amp[\"Gene name\"])) | (vvx[\"Position\"].isin(vvx_kl[\"Position\"]))]\n",
    "\n",
    "#Test on those regions\n",
    "vvv = prepare_alpha(vx_use, vvx_use, stat = \"PhyloP447\")\n",
    "cuttt = 0.6\n",
    "z = list(vvx_amp[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Known hotspots of positive selection on chrX\")\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9150fe98",
   "metadata": {},
   "outputs": [],
   "source": [
    "vx_use = vx[vx[\"NearestGene\"].isin([\"LAS1L\"])]\n",
    "vvx_use = vvx[vvx[\"NearestGene\"].isin([\"LAS1L\"])]\n",
    "\n",
    "#Test on those regions\n",
    "vvv = prepare_alpha(vx_use, vvx_use, stat = \"PhyloP447\")\n",
    "cuttt = 0.9\n",
    "z = list(vvx_amp[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Known hotspots of positive selection on chrX\", window = [-5, 12])\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)\n",
    "\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44defb80",
   "metadata": {},
   "outputs": [],
   "source": [
    "vx_use = vx[(vx[\"NearestGene\"].isin(amp[\"Gene name\"])) | (vx[\"Position\"].isin(vx_kl[\"Position\"]))]\n",
    "vvx_use = vvx[(vvx[\"NearestGene\"].isin(amp[\"Gene name\"])) | (vvx[\"Position\"].isin(vvx_kl[\"Position\"]))]\n",
    "\n",
    "#Test on those regions\n",
    "vvv = prepare_alpha(vx_use, vvx_use, stat = \"PhyloP447\")\n",
    "cuttt = 0.6\n",
    "z = list(vvx_amp[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Known hotspots of positive selection on chrX\", window = [3, 12])\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)\n",
    "\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60b4d19f",
   "metadata": {},
   "outputs": [],
   "source": [
    "vx_use = vx[~((vx[\"NearestGene\"].isin(amp[\"Gene name\"])) | (vx[\"Position\"].isin(vx_kl[\"Position\"])))]\n",
    "vvx_use = vvx[~((vvx[\"NearestGene\"].isin(amp[\"Gene name\"])) | (vvx[\"Position\"].isin(vvx_kl[\"Position\"])))]\n",
    "\n",
    "#Test on those regions\n",
    "vvv = prepare_alpha(vx_use, vvx_use, stat = \"PhyloP447\")\n",
    "cuttt = 0.6\n",
    "z = list(vvx_amp[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Outside known hotspots of positive selection on chrX\")\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "737089d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test on those regions\n",
    "vvv = prepare_alpha(vx_use, vvx_use, stat = \"PhyloP447\")\n",
    "cuttt = 0.6\n",
    "z = list(vvx_amp[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Known hotspots of positive selection on chrX\", window = [3, 12])\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)\n",
    "\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08bb7da2",
   "metadata": {},
   "outputs": [],
   "source": [
    "vx_use = vx[~((vx[\"NearestGene\"].isin(amp[\"Gene name\"])) | (vx[\"Position\"].isin(vx_kl[\"Position\"])))]\n",
    "vvx_use = vvx[~((vvx[\"NearestGene\"].isin(amp[\"Gene name\"])) | (vvx[\"Position\"].isin(vvx_kl[\"Position\"])))]\n",
    "\n",
    "#Test on those regions\n",
    "vvv = prepare_alpha(vx_use, vvx_use, stat = \"PhyloP447\")\n",
    "cuttt = 0.6\n",
    "z = list(vvx_amp[\"PhyloP447\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, cutoff = cutoff, title = \"Outside known hotspots of positive selection on chrX\", window = [3, 12])\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "print(alpha)\n",
    "\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2624ce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "hpo = pd.read_csv(\"../DPSC_CNCC/HPO_AccelEvol_Input.txt\", sep= \"\\t\")\n",
    "d_HPO = {}\n",
    "\n",
    "for index, row in hpo.iterrows():\n",
    "    d_HPO[row[\"Term\"]] = row[\"Genes\"].split(\";\")\n",
    "\n",
    "gobp = pd.read_csv(\"../DPSC_CNCC/GOBP_AccelEvol_Input.txt\", sep= \"\\t\")\n",
    "d_BP = {}\n",
    "\n",
    "for index, row in gobp.iterrows():\n",
    "    if \"ucus\" in row[\"Term\"]:\n",
    "        print(row[\"Term\"])\n",
    "    d_BP[row[\"Term\"]] = row[\"Genes\"].split(\";\")\n",
    "    \n",
    "kegg = pd.read_csv(\"../DPSC_CNCC/KEGG_AccelEvol_Input.txt\", sep= \"\\t\")\n",
    "d_KEGG = {}\n",
    "\n",
    "for index, row in kegg.iterrows():\n",
    "    if \"ucus\" in row[\"Term\"]:\n",
    "        print(row[\"Term\"])\n",
    "    d_KEGG[row[\"Term\"]] = row[\"Genes\"].split(\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35cd4524",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test per geneset on chrX\n",
    "c = 0\n",
    "out = []\n",
    "vx.index = vx[\"NearestGene\"]\n",
    "vvx.index = vvx[\"NearestGene\"]\n",
    "\n",
    "for key in d_HPO.keys():\n",
    "    if c%1000 == 0:\n",
    "        print(c)\n",
    "    if len(d_HPO[key]) >= 15 and len(d_HPO[key]) < 100:\n",
    "        vk = vx.loc[np.intersect1d(np.intersect1d(vx.index, d_HPO[key]), vvx.index)]\n",
    "        vvk = vvx.loc[np.intersect1d(np.intersect1d(vvx.index, d_HPO[key]), vx.index)]\n",
    "        x2 = [np.float64(x) for x in list(vk[\"PhyloP447\"])]\n",
    "        yvals2 = [np.float64(j) for j in list(vvk[\"PhyloP447\"])]\n",
    "        yvals2.sort()\n",
    "        if len(vk.index) >= 100 and len(vvk.index) >= 50:\n",
    "            vvv = prepare_alpha(vk, vvk, stat = \"PhyloP447\")\n",
    "            for cuttt in [0.6, 0.7, 0.8, 0.9]:\n",
    "                cutoff = yvals2[int(floor((len(yvals2)*cuttt)))]\n",
    "                alpha = compute_alpha_cutoff(vvv, dn_cut = 0.05, plot = False, cutoff = cutoff)\n",
    "\n",
    "                table1 = alpha[2]\n",
    "                table2 = alpha[3]\n",
    "                \n",
    "                out.append([key, np.median(x2), len(x2), np.median(yvals2), len(yvals2), (fisher_exact(table1)[1] + fisher_exact(table2)[1])/2, mwu(x2, yvals2)[1], (fisher_exact(table1, alternative = \"greater\")[1] + fisher_exact(table2, alternative = \"greater\")[1])/2, mwu(x2, yvals2, alternative = \"greater\")[1], alpha[0], alpha[1], table1, table2, cuttt])\n",
    "    c += 1\n",
    "    \n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "df = pd.DataFrame(out)\n",
    "df = df.sort_values(7)\n",
    "df.columns = [\"Term\", \"Median Fixed PhyloP\", \"Number Fixed Variants\", \"Median Polymorphic PhyloP\", \"Number Polymorphic Variants\", \"Fisher exact p-value\", \"MWU p-value\", \"Fisher exact p-value; alt greater\", \"MWU p-value; alt greater\", \"alpha\", \"Cutoff\", \"[[dc1, du1], [pc1, pu1]]\", \"[[dc2, du2], [pc2, pu2]]\", \"Proportion\"]\n",
    "df.to_csv(\"Finalizing_FisherExact_Unfold_NewHPO_NonCod_MafCut0.25-0.75_chrX_NonCod_SpecSup250.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9ee3e2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"ForPub_DoubleIntersect_FisherExact_Unfold_NewHPO_NonCod_MafCut0.25-0.75_chrX_NonCod_SpecSup250.csv\")\n",
    "new_df = pd.DataFrame()\n",
    "for prop in np.unique(df[\"Proportion\"]):\n",
    "    df2 = df[df[\"Proportion\"] == prop]\n",
    "    df2[\"Fisher exact FDR\"] = fdrcorrection(df2[\"Fisher exact p-value; alt greater\"])[1]\n",
    "    df2[\"MWU FDR\"] = fdrcorrection(df2[\"MWU p-value; alt greater\"])[1]\n",
    "    new_df = pd.concat([new_df, df2])\n",
    "df = new_df.sort_values(\"Fisher exact FDR\")\n",
    "df_sig = df[df[\"Fisher exact FDR\"] < 0.1]\n",
    "df_sig2 = df[df[\"MWU FDR\"] < 0.1]\n",
    "df_sig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f82f668",
   "metadata": {},
   "outputs": [],
   "source": [
    "key = \"Short toe\"\n",
    "\n",
    "vx.index = vx[\"NearestGene\"]\n",
    "vvx.index = vvx[\"NearestGene\"]\n",
    "\n",
    "vk = vx.loc[np.intersect1d(np.intersect1d(vx.index, d_HPO[key]), vvx.index)]\n",
    "vvk = vvx.loc[np.intersect1d(np.intersect1d(vvx.index, d_HPO[key]), vx.index)]\n",
    "x2 = [np.float64(x) for x in list(vk[\"PhyloP447\"])]\n",
    "yvals2 = [np.float64(j) for j in list(vvk[\"PhyloP447\"])]\n",
    "yvals2.sort()\n",
    "if len(vk.index) >= 100 and len(vvk.index) >= 50:\n",
    "    vvk = vvk[~vvk[\"NearestGene\"].isin([\"SHOX\"])]\n",
    "    vvv = prepare_alpha(vk, vvk, stat = \"PhyloP447\")\n",
    "    cuttt = 0.9\n",
    "    cutoff = yvals2[int(floor((len(yvals2)*cuttt)))]\n",
    "    alpha = compute_alpha_cutoff(vvv, dn_cut = 0.05, plot = True, cutoff = cutoff, title = key + \" genes on chrX\")\n",
    "\n",
    "    table1 = alpha[2]\n",
    "    table2 = alpha[3]\n",
    "    print((fisher_exact(table1, alternative = \"greater\")[1] + fisher_exact(table2, alternative = \"greater\")[1])/2)\n",
    "    print((fisher_exact(table1, alternative = \"greater\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "402826bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "vk = vx.loc[np.intersect1d(vx.index, [\"GPC4\"])]\n",
    "vvk = vvx.loc[np.intersect1d(vvx.index, [\"GPC4\"])]\n",
    "x2 = [np.float64(x) for x in list(vk[\"PhyloP447\"])]\n",
    "yvals2 = [np.float64(j) for j in list(vvk[\"PhyloP447\"])]\n",
    "yvals2.sort()\n",
    "vvv = prepare_alpha(vk, vvk, stat = \"PhyloP447\")\n",
    "cuttt = 0.9\n",
    "cutoff = yvals2[int(floor((len(yvals2)*cuttt)))]\n",
    "alpha = compute_alpha_cutoff(vvv, dn_cut = 0.05, plot = True, cutoff = cutoff, title = \"$\\it{GPC4}$\")\n",
    "\n",
    "table1 = alpha[2]\n",
    "table2 = alpha[3]\n",
    "print((fisher_exact(table1, alternative = \"greater\")[1] + fisher_exact(table2, alternative = \"greater\")[1])/2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
