{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "394936a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PosSelect_Functions_Old import *\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import gaussian_kde\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy\n",
    "import seaborn as sns\n",
    "from scipy.stats import mannwhitneyu as mwu\n",
    "from scipy.stats import ttest_ind\n",
    "from scipy.stats import ttest_rel\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from scipy.stats import wilcoxon\n",
    "from scipy.optimize import curve_fit\n",
    "from scipy.stats import fisher_exact\n",
    "from scipy.stats import norm\n",
    "from collections import Counter\n",
    "from scipy.stats import binomtest\n",
    "from scipy.stats import spearmanr,pearsonr\n",
    "\n",
    "hfont = {'fontname':'Arial'}\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "\n",
    "#Code borrowed heavily from here: https://stackoverflow.com/questions/62375034/find-non-overlapping-area-between-two-kde-plots\n",
    "plt.rcParams.update(\n",
    "    {\"text.usetex\": False}\n",
    ")\n",
    "\n",
    "def parse_table(i):\n",
    "    return [int(x) for x in i.replace(\"[\", \"\").replace(\"]\", \"\").split(\",\")]\n",
    "\n",
    "d_abrev = {\"LiangSteinNeuron\":\"FC exc. neur.\", \"FetalChondrocytes\":\"F chond.\", \"SertoliMale\":\"FG sertoli\", \"preGC_IIaFemale\":\"FG preGC IIa\",\\\n",
    "          \"NeuralFemale\":\"FG neur.\", \"FetalGonadImmuneFemale\":\"FG immune\", \"VIP\":\"AC VIP inh. neur.\", \"LiangSteinProgenitor\":\"FC prog.\",\\\n",
    "          \"AdultHeartVentricularCardiomyocyte\":\"AH cardiomyo.\", \"AdultLoopOfHenle\":\"AK loop of henle\", \"FetalBrainNeurGlioblast_CB_VZ\":\"FCB glioblast\",\\\n",
    "         \"AdultProximalTubule\":\"AK prox. tub.\", \"FetalLeydigMale\":\"FG leydig\", \"SST\":\"AC SST inh neur.\", \"KosoyRoussosControlMicroglia\":\"AC microglia\",\\\n",
    "         \"FetalBrainFloorPlate\":\"FB fl. plate\", \"FetalArterialECs\":\"FH endoth.\", \"ASCT\":\"AC astro.\", \"FetalBrainCOP\":\"FB COP\",\\\n",
    "         \"AMY\":\"AA neur.\", \"PVALB\":\"AC PVALB inh neur.\", \"ITL23\":\"AC L2-3 IT neur.\", \"FetalBrainNeurCB_GNP_IPC_1\":\"FB inter. prog.\", \"FetalBrainNeurDAergic\":\"FB DA neur.\",\\\n",
    "          \"OGC\":\"AC Oligo.\", \"D1Pu\":\"AP D1 inh neur.\", \"FetalBrainNeurSerotonergic\":\"FB 5-HT neur.\", \"FetalBrainNeurDRG_2\":\"FS DRG neur.\",\\\n",
    "          \"FetalHeartPericytes\":\"FH peri.\", \"FetalHeartEndocardium\":\"FH endocard.\", \"FetalHeartCardiacFibroblasts\":\"FH fibro.\", \"FetalBrainNeurPurkinje_6\":\"FCB Purk. inh neur.\",\\\n",
    "          \"AdultHeartSmoothMuscle\":\"AH smooth musc.\", \"FetalBrainRoofPlate\":\"FB ro. plate\"}\n",
    "\n",
    "gocc = pd.read_csv(\"../DPSC_CNCC/GOCC_AccelEvol_Input.txt\", sep= \"\\t\")\n",
    "d_CC = {}\n",
    "\n",
    "for index, row in gocc.iterrows():\n",
    "    d_CC[row[\"Term\"]] = row[\"Genes\"].split(\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ed747bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "v2 = pd.read_csv(\"TF_Data/MA1643.1_NFIB_400_LiangSteinNeuron_Fixed.csv.gz\")\n",
    "vv2 = pd.read_csv(\"TF_Data/MA1643.1_NFIB_400_LiangSteinNeuron_Poly.csv.gz\")\n",
    "v2 = v2[(v2[\"1\"] > 500) | ((v2[\"2\"] > 500) & (v2[\"4\"] > 500))]\n",
    "vv2 = vv2[(vv2[\"1\"] > 500) | ((vv2[\"2\"] > 500) & (vv2[\"3\"] > 500))]\n",
    "\n",
    "be = list(pd.read_csv(\"BlacklistTE_Variants.txt\")[\"Position\"])\n",
    "v2 = v2[~v2[\"Position\"].isin(be)]\n",
    "vv2 = vv2[~vv2[\"Position\"].isin(be)]\n",
    "\n",
    "v2_fore = v2[v2[\"abs logfc\"] > 0.25]\n",
    "\n",
    "v2_fore[\"Chrom\"] = [x.split(\":\")[0] for x in v2_fore[\"Position\"]]\n",
    "v2_fore[\"Pos1\"] = [int(x.split(\":\")[1]) for x in v2_fore[\"Position\"]]\n",
    "v2_fore[\"Pos2\"] = [int(x.split(\":\")[1]) + 1 for x in v2_fore[\"Position\"]]\n",
    "v2_fore = v2_fore[[\"Chrom\", \"Pos1\", \"Pos2\"]]\n",
    "\n",
    "#Need to change \n",
    "v2_fore.to_csv(\"TCF12_AbsG0.25_Fixed_LSN_FilterNewTEs.bed\", sep = \"\\t\", header = None, index = None)\n",
    "\n",
    "v2[\"Chrom\"] = [x.split(\":\")[0] for x in v2[\"Position\"]]\n",
    "v2[\"Pos1\"] = [int(x.split(\":\")[1]) for x in v2[\"Position\"]]\n",
    "v2[\"Pos2\"] = [int(x.split(\":\")[1]) + 1 for x in v2[\"Position\"]]\n",
    "\n",
    "#Set to 0.05 for MEF2A\n",
    "#Change 0.025 to 0.1 for NFIB and other TFs with large effects on CA\n",
    "v2_back = pd.concat([v2_fore, v2[v2[\"abs logfc\"] < 0.1]])\n",
    "v2_back = v2_back[[\"Chrom\", \"Pos1\", \"Pos2\"]]\n",
    "v2_back.to_csv(\"TCF12_AbsG0.25_Fixed_Background_New_preGC_FilterNewTEs.bed\", sep = \"\\t\", header = None, index = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec5eaf6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Can change various things in this cell to make plots for different TFs, restrict to synaptic genes, etc.\n",
    "#Enrichment for Rxra gets even stronger at 550, 600\n",
    "#Plot a TF by changing the ID\n",
    "\n",
    "v2 = pd.read_csv(\"TF_Data/MA1643.1_NFIB_400_LiangSteinNeuron_Fixed.csv.gz\")\n",
    "vv2 = pd.read_csv(\"TF_Data/MA1643.1_NFIB_400_LiangSteinNeuron_Poly.csv.gz\")\n",
    "v2 = v2[(v2[\"1\"] > 500) | ((v2[\"2\"] > 500) & (v2[\"4\"] > 500))]\n",
    "vv2 = vv2[(vv2[\"1\"] > 500) | ((vv2[\"2\"] > 500) & (vv2[\"3\"] > 500))]\n",
    "\n",
    "be = list(pd.read_csv(\"BlacklistTE_Variants.txt\")[\"Position\"])\n",
    "v2 = v2[~v2[\"Position\"].isin(be)]\n",
    "vv2 = vv2[~vv2[\"Position\"].isin(be)]\n",
    "\n",
    "#Can comment or uncomment to change synaptic restriction\n",
    "syngo = pd.read_csv(\"../Cell_Type_Prop/Cell_Type_Prop/syngo_genes.csv\")\n",
    "syngo = syngo[\"hgnc_symbol\"]\n",
    "\n",
    "#v2 = v2[~v2[\"NearestGene\"].isin(syngo)]\n",
    "#vv2 = vv2[~vv2[\"NearestGene\"].isin(syngo)]\n",
    "\n",
    "#v2 = v2[v2[\"abs logfc\"] > 0.25]\n",
    "#vv2 = vv2[vv2[\"abs logfc\"] > 0.25]\n",
    "\n",
    "#v2 = v2[v2[\"SpecSup447\"] > 250]\n",
    "#vv2 = vv2[vv2[\"SpecSup447\"] > 250]\n",
    "\n",
    "vv2 = vv2[vv2[\"Human ref\"] != vv2[\"Chimp ref\"]]\n",
    "cuttt = 0.6\n",
    "\n",
    "z = list(vv2[\"abs logfc\"])\n",
    "z.sort()\n",
    "cutoff = z[int(floor((len(z)*cuttt)))]\n",
    "\n",
    "vvv = prepare_alpha(v2, vv2, stat = \"abs logfc\")\n",
    "#vvv[\"PhyloP\"] = -vvv[\"PhyloP\"]\n",
    "alpha = compute_alpha_cutoff(vvv, plot = True, window = [0.2, 0.8], cutoff = cutoff, title = \"NFIB binding sites in fetal cortical exc. neurons\")\n",
    "plt.xlabel(\"Predicted absolute log$_{2}$ fold-change in CA\", fontsize = 20)\n",
    "plt.ylabel(\"Density\", fontsize = 20)\n",
    "print(alpha)\n",
    "print((fisher_exact(alpha[-2], alternative = \"greater\")[1] + fisher_exact(alpha[-3], alternative = \"greater\")[1])/2)\n",
    "\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "#plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a16bb46a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Across various cutoffs, we observe that when the sign of the motif difference is consistent with the sign of the logfc\n",
    "#Is consistent with the sign of the RNA log fold-change, we observe a significant bias toward downregulation of the gene\n",
    "#This holds across cutoffs for abs logfc, padj for RNA, and Top20th\n",
    "cutoff_lsn = (531.89703 + 520.9438)/2\n",
    "\n",
    "v2ps = v2p[v2p[\"abs logfc\"] > 0.1]\n",
    "x = pd.read_csv(\"Agoglia_Fraser_2021_DESeq2_HumChp_Hybrid_CS_D100_chpr.txt\", sep = \"\\t\").dropna()\n",
    "x = x[x[\"padj\"] < 0.05]\n",
    "x = x[x[\"Gene\"].isin(v2p[\"NearestGene\"])]\n",
    "\n",
    "v2ps[\"Sign1\"] = np.sign(v2ps[\"logfc\"])\n",
    "v2ps[\"Dif\"] = v2ps[\"1\"] - v2ps[\"2\"]\n",
    "v2ps[\"Sign2\"] = np.sign(v2ps[\"Dif\"])\n",
    "v2ps[\"Sign\"] = v2ps[\"Sign1\"]*v2ps[\"Sign2\"]\n",
    "v2ps = v2ps[v2ps[\"Sign\"] < 0]\n",
    "\n",
    "v2ps = v2ps[v2ps[\"NearestGene\"].isin(x[\"Gene\"])]\n",
    "v2ps = v2ps.set_index(\"NearestGene\").join(x.set_index(\"Gene\"))\n",
    "v2ps = v2ps[(v2ps[\"allele1_pred_counts\"] > cutoff_lsn) | (v2ps[\"allele2_pred_counts\"] > cutoff_lsn)]\n",
    "v2ps[\"Sign3\"] = np.sign(v2ps[\"log2FoldChange\"])\n",
    "v2ps[\"Sign4\"] = v2ps[\"Sign3\"]*v2ps[\"Sign1\"]\n",
    "\n",
    "#Testing whether there is more agreement in NFIB effect and RNA l2fc than expected by chance\n",
    "agree = v2ps[v2ps[\"Sign4\"] < 0].shape[0]\n",
    "disagree = v2ps[v2ps[\"Sign4\"] > 0].shape[0]\n",
    "print(binomtest(agree, agree + disagree))\n",
    "\n",
    "v2ps = v2ps[v2ps[\"Sign4\"] < 0]\n",
    "v2ps_nd = v2ps.drop_duplicates(\"pvalue\")\n",
    "down = v2ps_nd[v2ps_nd[\"log2FoldChange\"] < 0].shape[0]\n",
    "up = v2ps_nd[v2ps_nd[\"log2FoldChange\"] > 0].shape[0]\n",
    "print(binomtest(down, down + up))\n",
    "\n",
    "#down = v2ps[v2ps[\"log2FoldChange\"] < 0].shape[0]\n",
    "#up = v2ps[v2ps[\"log2FoldChange\"] > 0].shape[0]\n",
    "#print(binomtest(down, down + up))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "872536af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#barplot of ASE values\n",
    "to_plot = v2ps_nd.sort_values(\"log2FoldChange\")\n",
    "out = []\n",
    "for index, row in to_plot.iterrows():\n",
    "    #out.append([\"$_{\" + index + \"}$\", row[\"log2FoldChange HumChp\"], \"Hybrid cortical organoid\"])\n",
    "    #out.append([\"$_{\" + index + \"}$\", row[\"log2FoldChange HumChp Adult\"], \"Adult MTG L2/3 IT\"])\n",
    "    if row[\"log2FoldChange\"] < 0:\n",
    "        x = \"Down human\"\n",
    "    else:\n",
    "        x = \"Up human\"\n",
    "    out.append([\"$_{\" + index + \"}$\", row[\"log2FoldChange\"], x])\n",
    "to_plot = pd.DataFrame(out)\n",
    "to_plot.columns = [\"Gene\", \"Log$_{2}$ fold-change\", \"Direction\"]\n",
    "sns.set(font_scale = 1.5)\n",
    "sns.set_style(\"white\")\n",
    "fig, ax = plt.subplots(figsize = (9, 4))\n",
    "sns.set_style(\"white\")\n",
    "new_palette = {\"Human\":\"#FF2C0C\", \"Chimp\":\"#0058FF\"}\n",
    "t_ax = sns.barplot(data = to_plot, x = \"Gene\", y = \"Log$_{2}$ fold-change\", hue = \"Direction\")\n",
    "\n",
    "\n",
    "c = 0\n",
    "for patch in t_ax.patches:\n",
    "    if c < 23:\n",
    "        patch.set_edgecolor(new_palette[\"Chimp\"])\n",
    "        patch.set_facecolor(new_palette[\"Chimp\"] + \"1A\")\n",
    "    else:\n",
    "        patch.set_edgecolor(new_palette[\"Human\"])\n",
    "        patch.set_facecolor(new_palette[\"Human\"] + \"1A\")\n",
    "    c += 1\n",
    "plt.title(\"ASE for synaptic genes near NFIB binding sites in day 100 cortical organoids\", size = 15)\n",
    "plt.xticks(rotation=90)\n",
    "plt.ylabel(\"Log$_{2}$ fold-change human/chimp\", size = 14)\n",
    "plt.legend([], [], frameon = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bea16201",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Organizing for biorender plot using SYNGO annotations + looking up on Uniprot\n",
    "local = {\"Presynapse\":[\"ADORA2B\", \"CACNA2D3\", \"CADM1\", \"CD47\", \"CPEB2\", \"CTNNB1\", \"GRIN2A\", \"OTOF\", \"PCDH9\", \"PRKCA\", \"PTPRN2\", \"PTPRS\", \"RIMBP2\", \"RPL13A\", \"RPL38\", \"STX7\", \"SV2C\", \"VLDLR\", \"VPS35\"], \"Postsynapse\":[\"ARF1\", \"ARHGAP32\", \"CADM1\", \"CPEB2\", \"CTNNB1\", \"GRIN2A\", \"LASP1\", \"PCDH9\", \"RAB4A\", \"RPL13A\", \"RPL38\", \"SIPA1L3\", \"SRGAP2\", \"VLDLR\", \"VPS35\"], \"Membrane\":[\"CDH23\", \"FKBP1A\", \"NCAM2\", \"RAB3C\"]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb6eb581",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in local[\"Membrane\"]:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb5deb23",
   "metadata": {},
   "outputs": [],
   "source": [
    "v2ps_nd.loc[local[\"Membrane\"]][v2ps_nd[\"log2FoldChange\"] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02df30ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "syn = pd.read_csv(\"syngo_annotations.csv\")\n",
    "syn[syn[\"hgnc_symbol\"].isin(v2ps_nd.index)].tail(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "067cb914",
   "metadata": {},
   "outputs": [],
   "source": [
    "v2ps_nd.loc[local[\"Postsynapse\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3734936e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Barplot in cortical organoids\n",
    "new_palette = {\"Human\":\"#FF2C0C\", \"Chimp\":\"#0058FF\"}\n",
    "\n",
    "sns.set(font_scale = 1.4)\n",
    "sns.set_style(\"white\")\n",
    "fig, ax = plt.subplots(figsize=(3.5,5))\n",
    "t_ax = sns.barplot({\"Upregulated\\nin human\":7, \"Downregulated\\nin human\":23}, errorbar=None, linewidth=2.5, edgecolor=\".5\", facecolor='#F2C91140', gap = 0.1, palette = {\"Upregulated\\nin human\":new_palette[\"Human\"], \"Downregulated\\nin human\":new_palette[\"Chimp\"]})\n",
    "for patch in t_ax.patches:\n",
    "    print(patch.get_x() + patch.get_width()/2)\n",
    "c = 0\n",
    "for patch in t_ax.patches:\n",
    "    if c < 1:\n",
    "        patch.set_edgecolor(new_palette[\"Human\"])\n",
    "        patch.set_facecolor(new_palette[\"Human\"] + \"1A\")\n",
    "    elif c == 1:\n",
    "        patch.set_edgecolor(new_palette[\"Chimp\"])\n",
    "        patch.set_facecolor(new_palette[\"Chimp\"] + \"1A\")\n",
    "    c += 1\n",
    "plt.ylabel(\"Number of genes\")\n",
    "plt.title(\"Day 100 cortical organoids\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bf06e94",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Permutation test to confirm GREAT results\n",
    "real = np.mean(v2p[\"abs logfc\"])\n",
    "size = v2p.shape[0]\n",
    "\n",
    "fakes = []\n",
    "for i in range(10000):\n",
    "    np.random.seed(i)\n",
    "    fake = np.mean(v2.sample(n = size, replace = False)[\"abs logfc\"])\n",
    "    fakes.append(fake)\n",
    "print(norm.sf((real - np.mean(fakes))/np.std(fakes)))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c77605d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.kdeplot(fakes, fill = True, color = \"#40A94D\")\n",
    "plt.vlines(x = real, ymin = 0, ymax = 100, color = \"#F42FF5\")\n",
    "plt.ylim(0, 100)\n",
    "plt.xlabel(\"Mean abs logfc of gene set\", size = 14)\n",
    "plt.ylabel(\"Density of random subsets\", size = 14)\n",
    "plt.title(\"MEF2A binding sites near synaptic assembly regulators\", size = 18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0274cf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create inputs to GREAT to test whether sites with large abs logfc are enriched near certain types of genes\n",
    "\n",
    "#Change 0.1 to 0.25 for NFIB and other TFs with large effects on CA\n",
    "v2_fore = v2[v2[\"abs logfc\"] > 0.1]\n",
    "\n",
    "v2_fore[\"Chrom\"] = [x.split(\":\")[0] for x in v2_fore[\"Position\"]]\n",
    "v2_fore[\"Pos1\"] = [int(x.split(\":\")[1]) for x in v2_fore[\"Position\"]]\n",
    "v2_fore[\"Pos2\"] = [int(x.split(\":\")[1]) + 1 for x in v2_fore[\"Position\"]]\n",
    "v2_fore = v2_fore[[\"Chrom\", \"Pos1\", \"Pos2\"]]\n",
    "\n",
    "#Need to change \n",
    "v2_fore.to_csv(\"ESR1_AbsG0.1_Fixed_preGC_FilterNewTEs.bed\", sep = \"\\t\", header = None, index = None)\n",
    "\n",
    "v2[\"Chrom\"] = [x.split(\":\")[0] for x in v2[\"Position\"]]\n",
    "v2[\"Pos1\"] = [int(x.split(\":\")[1]) for x in v2[\"Position\"]]\n",
    "v2[\"Pos2\"] = [int(x.split(\":\")[1]) + 1 for x in v2[\"Position\"]]\n",
    "\n",
    "#Set to 0.05 for MEF2A\n",
    "#Change 0.025 to 0.1 for NFIB and other TFs with large effects on CA\n",
    "v2_back = pd.concat([v2_fore, v2[v2[\"abs logfc\"] < 0.025]])\n",
    "v2_back = v2_back[[\"Chrom\", \"Pos1\", \"Pos2\"]]\n",
    "v2_back.to_csv(\"ESR1_AbsG0.1_Fixed_Background_New_preGC_FilterNewTEs.bed\", sep = \"\\t\", header = None, index = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f2cd10e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### COPY OF CELL ABOVE BUT RECORDING WHAT WE DID FOR NFIB ###\n",
    "\n",
    "#v2 = v2[(v2[\"allele1_pred_counts\"] > 350) | (v2[\"allele1_pred_counts\"] > 350)]\n",
    "v2_fore = v2[v2[\"abs logfc\"] > 0.25]\n",
    "\n",
    "v2_fore[\"Chrom\"] = [x.split(\":\")[0] for x in v2_fore[\"Position\"]]\n",
    "v2_fore[\"Pos1\"] = [int(x.split(\":\")[1]) for x in v2_fore[\"Position\"]]\n",
    "v2_fore[\"Pos2\"] = [int(x.split(\":\")[1]) + 1 for x in v2_fore[\"Position\"]]\n",
    "v2_fore = v2_fore[[\"Chrom\", \"Pos1\", \"Pos2\"]]\n",
    "v2_fore.to_csv(\"NFIB_AbsG0.25_Fixed_LSN_FilterNewTEs.bed\", sep = \"\\t\", header = None, index = None)\n",
    "\n",
    "v2[\"Chrom\"] = [x.split(\":\")[0] for x in v2[\"Position\"]]\n",
    "v2[\"Pos1\"] = [int(x.split(\":\")[1]) for x in v2[\"Position\"]]\n",
    "v2[\"Pos2\"] = [int(x.split(\":\")[1]) + 1 for x in v2[\"Position\"]]\n",
    "\n",
    "#Set to 0.05 for MEF2A\n",
    "v2_back = pd.concat([v2_fore, v2[v2[\"abs logfc\"] < 0.1]])\n",
    "v2_back = v2_back[[\"Chrom\", \"Pos1\", \"Pos2\"]]\n",
    "v2_back.to_csv(\"NFIB_AbsG0.25_Fixed_Background_New_LSN_FilterNewTEs.bed\", sep = \"\\t\", header = None, index = None)\n",
    "\n",
    "vv2[\"Chrom\"] = [x.split(\":\")[0] for x in vv2[\"Position\"]]\n",
    "vv2[\"Pos1\"] = [int(x.split(\":\")[1]) for x in vv2[\"Position\"]]\n",
    "vv2[\"Pos2\"] = [int(x.split(\":\")[1]) + 1 for x in vv2[\"Position\"]]\n",
    "v2_back = pd.concat([v2_fore, vv2])\n",
    "v2_back = v2_back[[\"Chrom\", \"Pos1\", \"Pos2\"]]\n",
    "v2_back.to_csv(\"NFIB_AbsG0.25_Fixed_Background_Poly_LSN_FilterNewTEs.bed\", sep = \"\\t\", header = None, index = None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
